### PR TITLE
net: sockets: dtls: reset mbedtls session on timed out handshake

### LINKED ISSUE #35711: net: sockets: dtls: handshake not reset as it ought to be
**Describe the bug**
The bug is as follows: I am a client in the DTLS handshake and the remote is not replying. I am using `sendto` to start the DTLS handshake. My end will retransmit the client hello message until timeout. This works well.

I will receive error on `sendto` since remote is gone. When I call0 `sendto` again, I would expect the DTLS handshake to be reset and start over if it's not complete. But what happens is that nothing comes out of the underlying socket because the mbedTLS session is not reset as it ought to be.

This happens:
1. sockets_tls.c calls `mbedtls_ssl_handshake` and gets back `MBEDTLS_ERR_SSL_TIMEOUT`. We return error and continue on our day. According to mbedTLS manual (paraphrased)
>If this function returns `MBEDTLS_ERR_SSL_TIMEOUT`, you must either free it or call mbedtls_ssl_session_reset() on it before re-using it for a new connection
https://tls.mbed.org/api/ssl_8h.html#a4a37e497cd08c896870a42b1b618186e

2. When I call `sendto` again, `mbedtls_ssl_handshake` will be called without being reset, and the internal state within mbedTLS will be "I'm waiting on data" -> nothing comes out and handshake is not retried.

Am I supposed to call close() if sendto returns error? This doesn't make any sense to me since I am operating on an UDP socket (not really, but I as a user would at least expect it to work in a similar way).

Is my expectations  wrong?

**To Reproduce**
1. Call sendto twice without remote.
2. On second call no client hello will come out.

**Expected behavior**
A new handshake should be started.

**Impact**
Annoyance if I have to call `close()` every time handshake fails.

**Environment (please complete the following information):**
 - Commit SHA 6e2522d

**Additional context**

### PR DESCRIPTION
**Pull request by emillindq**
This PR fixes #35711

According to MbedTLS API documentation, its session must be reset if mbedtls_ssl_handshake returns timeout error. This commit resets the session for said return value, and that allows us to call send() multiple times even if handshake times out for previous calls.

Please note that `k_sem_init` resets the semaphores internal list of polls waiting for it, and therefore a poll waiting on the `tls_established` sem will block forever if first handshake fails. This was changed to `k_sem_reset` in this PR.

This PR is a draft because I wanted to ask @rlubos why the `context->dtls_peer_addr` is erased in `tls_mbedtls_reset`? This breaks stuff since we then forget which our peer is, which is required if calling first `connect` and then `send`, [dtls hs timeout], `send` (which the echo_server/client does). What I mean is, is this strictly required in some cases? Thanks!

### CODE REVIEW

**Comment by rlubos** [comment]
> This PR is a draft because I wanted to ask @rlubos why the context->dtls_peer_addr is erased in tls_mbedtls_reset? This breaks stuff since we then forget which our peer is, which is required if calling first connect and then send, [dtls hs timeout], send (which the echo_server/client does). What I mean is, is this strictly required in some cases? Thanks!

This is needed for the DTLS server, where you don't `connect()` explicitly. Once some client establishes connection with a DTLS server, `dtls_peer_addr` is set to its address, so that we don't process packets from different peers (we only support a single DTLS session per socket). After the client ends the session, we reset the context **and** the address so that a new one can connect w/o a need to reopen a socket.

Since it shouldn't be needed for a client to clear the peer address, I guess it should be enough to verify the DTLS role in the `tls_mbedtls_reset()` function.

**Comment by emillindq** [comment]
> Since it shouldn't be needed for a client to clear the peer address, I guess it should be enough to verify the DTLS role in the `tls_mbedtls_reset()` function.

Thanks for this feedback; I have changed and added a comment about this.

I do feel quite confident in this PR; I have tested it with the following setup
1. Two stm32h747i_disco_m7 devkits: client and server
2. Hold button on server on boot to make it enter the network (aka. reply to ARP) but not reply to UDP/TCP.
3. Client and server is started with server put into mode above. TLS will fail and client will fail DTLS handshake over and over.
4. After a while server is rebooted with normal mode
5. We watch DTLS handshake succeed and echoing begins on UDP. TLS is not recovered since the connect() TCP call failed.

Tested on both IPV4 & IPV6. Please see https://github.com/emillindq/zephyr/tree/tls_handshake_reset_test for source.

**Comment by rlubos** [comment]
> 5\. We watch DTLS handshake succeed and echoing begins on UDP. TLS is not recovered since the connect() TCP call failed.

And I think that's what should be expected, right? The man pages for `connect()` are clear, that if the `connect()` call fails, one should close/reopen the socket before another attempt:

```
       If connect() fails, the state of the socket is unspecified.
       Conforming applications should close the file descriptor and
       create a new socket before attempting to reconnect.
```

**Review by rlubos — COMMENTED** [review]
Looks good, just one nit.

File: subsys/net/lib/sockets/sockets_tls.c [hunk 1]
```diff
@@ -809,12 +809,19 @@ static int tls_mbedtls_reset(struct tls_context *context)
...
-		     sizeof(context->dtls_peer_addr));
-	context->dtls_peer_addrlen = 0;
+	/* Server role: reset the address so that a new
+	 *              client can connect w/o a need to reopen a socket
+	 * Client role: keep peer addr so socket can continue to be used
+	 *              even on handshake timeout
+	 */
+	if (context->options.role == MBEDTLS_SSL_IS_SERVER) {
+		(void)memset(&context->dtls_peer_addr, 0,
+		sizeof(context->dtls_peer_addr));
```
[the remark this answers is not in the document]
Comment (emillindq) [reply subsys/net/lib/sockets/sockets_tls.c]: shoot, that was a function!

**Review by rlubos — APPROVED** [review]
LGMT, thanks!

**Review by cfriedt — APPROVED** [review]

**Pull request merged**